### ISSUE TITLE
Feature Request: Support for Flexible Font Weights (100 - 900) in Syntax Highlighting (and more css-properties)

### ISSUE THREAD

**Issue by iDad5** [microsoft/language-server-protocol#1545]
[Please excuse me if I'm in the wrong place or am doing anything else wrong - I'm rather new to this...]

I guess it won't be easy, but I really would appreciate the possibility to add font weights other than **bold** in syntax highlighting (and other theming-contexts as well).

Like most other developers I spent countless hours in front of my code, and having good typography is a lot more than just 'eye-candy'....

Luckily there are a lot of great developer friendly fonts to have nowadays and a lot of them come in various weights - some even variable. Syntax highlighting is a great boost to productivity, and being able to convey information that way is important.
For ergonomic reasons however I find it hard to work with too bright colors and when using less saturated colors it is often problematic to find a good balance, as yellow(ish) colors almost ever shine very bright, while (darker) reds for example tend to look very 'thin'.

I would really like to have the option to use font-weight 200 for yellows and 300 for reds for example. 

Also support for 'oblique' in addition to 'italic' would be great and other typographic css features too. 
I guess that TextMate-Syntax is a hard limiting factor, but maybe **semantic highlighting** could offer more of a choice?

Thank You!

BTW.: I do very much appreciate what you are giving us with VS-Code and would like to generally tank everyone working on that project!

**Comment by RedCMD** [comment]
Vscode currently only supports
normal
**bold**
_underline_
~~strikethrough~~

sadly you can't even combine them
it's either one or the other (at least in textmate)

would be nice to have other customizable styles

**Comment by iDad5** [comment]
As I'm getting more and more confused with every step I take with my VS-Code journey, I feel insecure even when I'm so sure that I have used combinations of italic and bold and strikethrough successfully....
;-)

**Comment by radeksimko** [comment]
To my knowledge, semantic token based highlighting - or even the static grammar based ones (such as TextMate used by VS Code) communicate primarily _meanings_ of tokens, not _how to highlight_ them.

e.g. LSP can say that in `foo = bar`, `foo` is an "attribute" and then it leaves it up to clients, how exactly to render attributes. It doesn't know or care whether that's red, blue, bold etc. - that's up to the client (i.e. VS Code).

Additionally, many clients will have some kind of theme support, so it seem counter-productive to have language servers suggest exact styling as they don't know anything about the UI.

TL;DR this sounds more like a feature for VS Code [themes], not for LSP?

**Comment by iDad5** [comment]
@radeksimko to be frank, I'm uncertain if you are deliberately misinterpreting my suggestion or if we have a genuine misunderstanding. 

Here's a clarification of my understanding:

As the name 'Syntax Highlighting' obviously gives away, my suggestion is about supporting advanced display features for highlighting. And I agree that it is the responsibility of themes to define the actual display style of different tokens.

Sadly, the way this works is not as clearly separated as it might be in an ideal world. 

Theme authors would likely love to have a good abstraction-layer like CSS-styles provide. What we have currently is only the possibility to define 'settings' for token-'scopes' in the form of 'foreground' (color) and 'fontStyle'.

The system is (as far as I understand it) based on TextMate and tries to stay largely compatible.
To my best understanding (and some confirmation I received in the past) if such a feature can be made available to theme authors it is only through LSP. This is not what one ideally would wish for but the way it is, and that isn't a situation related 
to VS-Code exclusively it seems.

A far better solution in my mind would be to write themes (at least the syntax highlighting) in CSS with token-scopes for selectors. After all, VS-Code and many other apps are more or less browsers at hart.

**Comment by puremourning** [comment]
What part of language server protocol would you actually like to be changed ? 

The semantic _tokens_ request returns tokens, and does not say _anything_ about styling, or rendering, etc. https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens

I think you probably need to request some change to VSCode, not the LSP.

**Comment by iDad5** [comment]
@puremourning I don't have the time and energy for a lengthy research in older threads and a comprehensive answer right now.
Teaching me about the concepts of LSP isn't the point here (I do understand the basics quite ok).
It appears that the possibilities for styling tokens through themes are very much dependent on what the LSP offers and how it communicates back to the host application. Key contributors to LSP have been heavily involved in making VS-Code default themes work with semantic tokens and my understanding was, and still is, that LSP is a key component for enabling the styling of tokens in addition to providing the semantics.
@dbaeumer likely would have not tagged my request as 'feature-request' and with 'semantic-tokens' right where it was made if it was totally unconnected to or misplaced in LSP—I guess.
With similar request at VS-Code, I was informed, that those would require changes to LSP…
In fact, I'm totally indifferent where my suggestion ends up, and I'm willing to discuss the pros and cons of my suggestion anytime. However, I don't see any sense in pure abstract discussions like where in heavily interconnected subjects a suggestion belongs, without providing a useful suggestion where to move it and how that would help.
If you can offer me an explanation of how my suggestion—even in theory—would work without changes to LSP I'm very much interested, though.

**Comment by puremourning** [comment]
I'm just trying to establish what you are requesting. If it amounts to:

> I would really like to have the option to use font-weight 200 for yellows and 300 for reds for example.

then that's a client thing. Your client (vscode, presumably) would need some configuration that said "render reds with 200 whatevers". The client can render the text however it likes, or however you like. It's not clear how LSP can/should be changed here. The LSP does not specify how clients render text.

**Comment by KamasamaK** [comment]
@iDad5 Styling tokens is not currently directly dependent on anything with LSP. That's why it was marked as a "feature request". 

That communication and abstraction between the language server and client is useless without the native client APIs. In this case, with VS Code you need to look at its native functionality. LSP will not adopt any new features without it _first_ being implemented in a client. VS Code does have native semantic tokens APIs, by the way, and the LSP semantic tokens API was largely _based on_ those -- see [DocumentSemanticTokensProvider](https://code.visualstudio.com/api/references/vscode-api#DocumentSemanticTokensProvider) and [DocumentRangeSemanticTokensProvider](https://code.visualstudio.com/api/references/vscode-api#DocumentRangeSemanticTokensProvider) for example.

**Comment by RedCMD** [comment]
iDad did state he was unsure how it should be implemented exactly

Both TextMate and semantic highlighting only describe what the token is and not how it should be styled

It is then up to the theme to stylise the tokens
obviously there are conventions to what certain token names are expected to do (`comment` disable intellisene and `markdown.itatlic` being italic etc)

I would think `fontWeight` would be a 3rd option in edition to `color` and `fontStyle` in themes
